### PR TITLE
Fix incorrect tag completion in code block

### PR DESCRIPTION
**Pull request by asmsuechan**
Finally, I fixed this problem which is reported by https://github.com/BoostIO/Boostnote/issues/333 and https://github.com/BoostIO/Boostnote/issues/294.

Before, this code didn't work well in CodeBlock: `<port protcol="tcp" port="10022"/>`

screen shot 2017-03-19 at 14 13 50

## before
screen shot 2017-03-19 at 14 15 29

## after
screen shot 2017-03-19 at 14 13 54

### CODE REVIEW

**Comment by asmsuechan** [comment]
And also I handled `<?php` issue reported by https://github.com/BoostIO/Boostnote/issues/275 and https://github.com/BoostIO/Boostnote/issues/205 by the second commit.

screen shot 2017-03-19 at 14 19 09

screen shot 2017-03-19 at 14 19 05

**Comment by asmsuechan** [comment]
Until the second commit, KaTeX didn't work well. But now after third commit, it works well.

screen shot 2017-03-19 at 15 26 09

screen shot 2017-03-19 at 15 26 05

File: browser/components/MarkdownPreview.js [hunk 1]
```diff
@@ -14,7 +14,8 @@ function decodeHTMLEntities (text) {
     ['apos', '\''],
     ['amp', '&'],
     ['lt', '<'],
-    ['gt', '>']
+    ['gt', '>'],
+    ['#63', '?']
   ]
 
   for (var i = 0, max = entities.length; i < max; ++i) {
```
Comment (asmsuechan) [inline browser/components/MarkdownPreview.js]: This `for` loop must be refactored.

**Comment by asmsuechan** [comment]
I fixed that bug, cut into a helper, and also added tests.

screen shot 2017-04-22 at 16 21 16

File: tests/lib/html-text-helper-test.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    '<a href=\'https://boostnote.io\'>Boostnote',
+    '<Component styleName=\'test\' />'
+  ]
+
+  testCases.forEach(testCase => {
+    const encodedText = htmlTextHelper.encodeEntities(testCase)
+    const decodedText = htmlTextHelper.decodeEntities(encodedText)
+    t.is(decodedText, testCase, 'Integration test through encodedText() and decodedText()')
+  })
+})
```
Comment (sota1235) [inline tests/lib/html-text-helper-test.js]: You are good JavaScript Engineer 👍

**Comment by kazup01** [comment]
o~

**Comment by kazup01** [comment]
fantastic👍

**Comment by PascalAOMS** [comment]
I'm running 0.11.12 and get this issue again.

Is there an option checkbox I need to tick or has the fix been broken with later updates?

**Pull request merged**